framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,588 @@
+     *   method call methodInvNode will be added to the parameter in the Scene.
+     *   If the Scene previously contained an annotated type for that parameter,
+     *   then its new type will be the LUB between the previous type and the
+     *   type of the respective value passed as argument in the method call.
+     *
+     * @param methodInvNode the node representing a method invocation.
+     * @param classSymbol the symbol of the class that contains the method being
+     * invoked.
+     * @param methodElt the element of the method being invoked.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the method parameters' types.
+     */
+    public static void updateMethodParameterTypeInScene(
+            MethodInvocationNode methodInvNode, ClassSymbol classSymbol,
+            ExecutableElement methodElt, AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return;
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString() +
+                ".jaif";
+        AScene scene = getScene(jaifPath);
+
+        AClass clazz = getJaifClass(classSymbol, scene);
+        if (clazz == null) return; // Anonymous class => Ignore, for now.
+        String methodName = JVMNames.getJVMMethodName(methodElt);
+        AMethod method = getJaifMethod(clazz, methodName, scene);
+        for (int i = 0; i < methodInvNode.getArguments().size(); i++) {
+            VariableElement ve = methodElt.getParameters().get(i);
+            if (atf.getAnnotatedType(ve).getExplicitAnnotations().size() > 0) {
+                // Ignore parameters that have a declared annotated type.
+                continue;
+            }
+
+            Node arg = methodInvNode.getArgument(i);
+            AnnotatedTypeMirror argType = atf.getAnnotatedType(arg.getTree());
+            AField param = method.parameters.vivify(i);
+            updateAnnotationSetInScene(
+                    param.type.tlAnnotationsHere, atf, jaifPath, argType);
+        }
+    }
+
+    /**
+     * Updates the type of the field lhs in the Scene of the class with
+     * symbol classSymbol.
+     *
+     * If the Scene contains no entry for the field lhs,
+     * the entry will be created and its type will be the type of rhs. If the
+     * Scene previously contained an entry/type for lhs, its new type will be
+     * the LUB between the previous type and the type of rhs.
+     *
+     * @param lhs the field.
+     * @param rhs the expression being assigned to the field.
+     * @param classSymbol the symbol of the class that contains the field.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the field's type.
+     */
+    public static void updateFieldTypeInScene(FieldAccessNode lhs, Node rhs,
+            ClassSymbol classSymbol, AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return;
```

Comment:
I expected the body of this method to be quite similar to that of updateMethodParameterTypeInScene just above, but the first few statements are in different orders.  Can you please choose a single structure that is preferable and use it in both bodies?  That will make both methods (and updateMethodReturnTypeInScene below) easier to understand.  Then, please see if it makes sense to abstract out some of the behavior, reducing code duplication.